pkg/ddc/jindo/metadata.go

```diff
@@ -1,5 +1,153 @@
+			"dataset namespace", dataset.Namespace,
+			"ufstotal", dataset.Status.UfsTotal)
+		should = false
+		return should, nil
+	}
+	should = true
+	return should, nil
+}
+
+func (e *JindoEngine) syncMetadataInternal() (err error) {
+	if e.MetadataSyncDoneCh != nil {
+		// Either get result from channel or timeout
+		select {
+		case result := <-e.MetadataSyncDoneCh:
+			defer func() {
+				e.MetadataSyncDoneCh = nil
+			}()
+			e.Log.Info("Get result from MetadataSyncDoneCh", "result", result)
+			if result.Done {
+				e.Log.Info("Metadata sync succeeded", "period", time.Since(result.StartTime))
+				err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+					dataset, err := utils.GetDataset(e.Client, e.name, e.namespace)
+					if err != nil {
+						return
+					}
+					datasetToUpdate := dataset.DeepCopy()
+					datasetToUpdate.Status.UfsTotal = result.UfsTotal
+					if !reflect.DeepEqual(datasetToUpdate, dataset) {
+						err = e.Client.Status().Update(context.TODO(), datasetToUpdate)
+						if err != nil {
+							return
+						}
+					}
+					return
+				})
+				if err != nil {
+					e.Log.Error(err, "Failed to update UfsTotal and FileNum of the dataset")
+					return err
+				}
+			} else {
+				e.Log.Error(result.Err, "Metadata sync failed")
+				return result.Err
+			}
+		case <-time.After(CHECK_METADATA_SYNC_DONE_TIMEOUT_MILLISEC * time.Millisecond):
+			e.Log.V(1).Info("Metadata sync still in progress")
+		}
+	} else {
+		// Metadata sync haven't started
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+			dataset, err := utils.GetDataset(e.Client, e.name, e.namespace)
+			if err != nil {
+				return
+			}
+			datasetToUpdate := dataset.DeepCopy()
+			datasetToUpdate.Status.UfsTotal = METADATA_SYNC_NOT_DONE_MSG
+			datasetToUpdate.Status.FileNum = METADATA_SYNC_NOT_DONE_MSG
+			if !reflect.DeepEqual(dataset, datasetToUpdate) {
+				err = e.Client.Status().Update(context.TODO(), datasetToUpdate)
+				if err != nil {
+					return
+				}
+			}
+			return
+		})
+		if err != nil {
+			e.Log.Error(err, "Failed to set UfsTotal to METADATA_SYNC_NOT_DONE_MSG")
+		}
+		e.MetadataSyncDoneCh = make(chan MetadataSyncResult)
+		go func(resultChan chan MetadataSyncResult) {
+			defer close(resultChan)
+			result := MetadataSyncResult{
+				StartTime: time.Now(),
+				UfsTotal:  "",
+			}
+			dataset, err := utils.GetDataset(e.Client, e.name, e.namespace)
+			if err != nil {
+				e.Log.Error(err, "Can't get dataset when syncing metadata", "name", e.name, "namespace", e.namespace)
+				result.Err = err
+				result.Done = false
+				resultChan <- result
+				return
+			}
+
+			e.Log.Info("Metadata Sync starts", "dataset namespace", e.namespace, "dataset name", e.name)
+
+			if len(dataset.Spec.Mounts) < 1 {
+				e.Log.Info("no dataset find")
+				result.Err = errors.New("Dataset Bind Error")
+			}
+			datasetUFSTotalBytes, err := e.TotalJindoStorageBytes(dataset.Spec.Mounts[0].Name)
```

Comment:
I noticed that `TotalJindoStorageBytes` will call `JindoFileUtils.GetUfsTotalSize` which is a func with timeout of 1500secs(i.e. 25min). Considering there may be a massive number of files to be synced and counted, is it better to change the func to a without-timeout one?

Reply:
As you said, it makes sense to change it this way